docs/extended_api/asynchronous_operations_library/memcpy_async.md

```diff
@@ -0,0 +1,64 @@
+---
+grand_parent: Extended API
+parent: Asynchronous operations library
+---
+
+# cuda::**memcpy_async**
+
+Defined in header [`<cuda/barrier>`](../../api/synchronization_library/barrier.md)
+
+```c++
+template<typename Shape, thread_scope Scope>
+void memcpy_async(void * destination, void const * source, Shape size, barrier<Scope> & barrier);                        // (1)
+
+template<typename Group, typename Shape, thread_scope Scope>
+void memcpy_async(Group const & group, void * destination, void const * source, Shape size, barrier<Scope> & barrier);   // (2)
+```
+
+Defined in header [`<cuda/pipeline>`](../headers/pipeline.md)
+
+```c++
+template<typename Shape, thread_scope Scope>
+void memcpy_async(void * destination, void const * source, Shape size, pipeline<Scope> & pipeline);                      // (3)
+
+template<typename Group, typename Shape, thread_scope Scope>
+void memcpy_async(Group const & group, void * destination, void const * source, Shape size, pipeline<Scope> & pipeline); // (4)
+```
+
+Asynchronously copies `size` bytes from the memory location pointed to by `source` to the memory location pointed to by `destination`.
+Both objects are reinterpreted as arrays of `unsigned char`.
+
+`cuda::memcpy_async` have similar constraints to [`std::memcpy`](https://en.cppreference.com/w/cpp/string/byte/memcpy), namely:
+* If the objects overlap, the behavior is undefined.
+* If either `destination` or `source` is an invalid or null pointer, the behavior is undefined (even if `count` is zero).
+* If the objects are potentially-overlapping or not [`TriviallyCopyable`](https://en.cppreference.com/w/cpp/named_req/TriviallyCopyable),
+  the behavior is undefined.

```

Comment:
If the objects are not `TriviallyCopyable` the behavior is not undefined.

Reply:
I think that the de jure `memcpy` is not specified to be undefined but may be so; making `memcpy_async` undefined per its spec is something I'm absolutely in favor of.

I believe that the entry points I wrote statically assert that the type is trivially copyable when possible. I believe that IFNDR would be a more modern way to spec it, but I'm unsure if the standard library uses that wording. @brycelelbach?

Reply:
The behavior of calling the following `well_defined` function is perfectly defined:

```c++
#include <algorithm>
#include <cstddef>
#include <cstring>
#include <type_traits>
struct NonTrivial { 
    NonTrivial() { }
};
static_assert(!std::is_trivial_v<NonTrivial>);

void well_defined() {
    NonTrivial a[2] = { NonTrivial{}, NonTrivial{} };
    std::aligned_storage_t<
      sizeof(NonTrivial[2]), 
      alignof(NonTrivial[2])
    > raw;
    std::memcpy(
        (std::byte*)&raw, 
        (std::byte*)a, 
        sizeof(NonTrivial[2])
    );
}
```

Reply:
The behavior of calling the following `well_defined` function is perfectly defined:

```c++
#include <algorithm>
#include <cstddef>
#include <cstring>
#include <type_traits>
struct NonTrivial { 
    NonTrivial() { }
};
static_assert(!std::is_trivial_v<NonTrivial>);

void well_defined() {
    NonTrivial a[2] = { NonTrivial{}, NonTrivial{} };
    std::aligned_storage_t<
      sizeof(NonTrivial[2]), 
      alignof(NonTrivial[2])
    > raw;
    std::memcpy(
        (std::byte*)&raw, 
        (std::byte*)a, 
        sizeof(NonTrivial[2])
    );
}
```

It just copies the bytes from `NonTrivial[2]` onto some uninitialized storage, which is fine. It never does anything that causes UB, like e.g., assuming that the objects at `raw` are live (they aren't, since we haven't called their constructors). 

A lot of code uses `memcpy` to copy raw bytes around like this (including bytes of non-trivial objects, uninitialized padding bytes, etc.). None of this is UB per se, and it is not UB for `memcpy`.

What's the rationale for making this "library UB" for `memcpy_async`?

Reply:
Trivial and trivially copyable are different properties. The type above should be trivially copyable, no?

Reply:
Non trivially copyable types are not guaranteed to be copyable with memcpy anywhere in the standard. See [basic.types.general] calling out trivially copyable types wherever it mentions memcpy.

Reply:
>  The type above should be trivially copyable, no?

It is. Adding a copy constructor makes it non-trivially copyable. I've fixed this above.

Reply:
Yes, and that makes it illegible for the additional allowances made by [basic.types.general], which is why it's UB here.

Non trivially copyable types are fragile when you copy them in a way different than by, well, copying them. I just noticed your code sample uses the power of std::byte to alias things, and it's invoking memcpy with a byte pointer, not a `NonTrivial` pointer, which means that it is copying the representation, not the object itself; that is fine and not prohibited by the call being UB if the objects are not trivially copyable. I admit the allowed cases of aliasing (through byte and chars) make this really muddy.

I tried to come up with a wording that makes this clearer, but I'm failing so far since the signatures don't talk about the actual pointer types, just about void pointers... But this follows what cppreference says about memcpy modulo turning "unspecified but maybe undefined" into "undefined", which IMHO is a much better place to be.

Reply:
> Non trivially copyable types are not guaranteed to be copyable with memcpy anywhere in the standard.

I disagree.

The standard guarantees that all objects have a representation, which consist of the sequence of sizeof(T) `unsigned char` objects taken up by the object (see [basic.types.general.4.1](http://eel.is/c++draft/basic.types.general#4.sentence-1)).

For `memcpy`, AFAICT the standard defers to the C standard library: [[cstring.syn.1]](http://eel.is/c++draft/cstring.syn#1), which just specifies that it copies N bytes, which are there, and has no notion of whether the types are trivially copyable or not, or C++ objects / lifetimes.

The `memcpy` above arguably creates new live objects of type `std::byte` in the `raw` storage; that's ok AFAICT. 

In [basic.types.general], the standard guarantees that for trivially copyable types, a round-trip `memcpy` to an array of `char`/`byte`/etc. preserves the object's _value_. It does not guarantee this anywhere for types that are not trivially copyable, but the example above does not use that.

Reply:
> I just noticed your code sample uses the power of std::byte to alias things, and it's invoking memcpy with a byte pointer, not a NonTrivial pointer

The signature of memcpy is `memcpy(void*, void*, size_t)`, not `memcpy(T*, T*, size_t)` (`memcpy` is not `std::copy`), so I'm not sure this matters. Whatever pointer type you take, there is a type conversion to `void*` that's going to happen anyways.

Reply:
It does matter; pointer values are the addresses in memory imbued with what object they point into.

Reply:
So? `memcpy` is just using that address to copy a representation.

Reply:
> I just noticed your code sample uses the power of std::byte to alias things,

Removing `std::byte` from the example still does not exhibit UB AFAICT:

```c++
#include <algorithm>
#include <cstddef>
#include <cstring>
#include <type_traits>
struct NonTrivial { NonTrivial() { } NonTrivial(NonTrivial const&) { }};
static_assert(!std::is_trivially_copyable_v<NonTrivial>);


void well_defined() {
    NonTrivial a[2] = { NonTrivial{}, NonTrivial{} };
    std::aligned_storage_t<
      sizeof(NonTrivial[2]), 
      alignof(NonTrivial[2])
    > raw;
    std::memcpy((void*)&raw, (void*)a, sizeof(NonTrivial[2]);
    );
}
```

Reply:
Taking a more pragmatic approach for a second, why would you **ever** want to do that (extracting a non-trivially-copyable object representation)? You cannot re-materialize a valid object from it.

Also, [[intro.object.8.4]](http://eel.is/c++draft/basic#intro.object-8.4) seems to indicate that:

> An object of trivially copyable or standard-layout type ([basic.types]) shall occupy contiguous bytes of storage.

I see no such requirement for non-trivially-copyable types, which seems incompatible with your claim.

Reply:
@gonzalobg I am not convinced that is well-defined; can you provide me with quotes of the normative text that define the behavior?

Reply:
> You cannot re-materialize a valid object from it.

Whether you can depends on the object (simple example: an object with a constructor that takes its raw representation can rematerialize itself from it).

> I see no such requirement for non-trivially-copyable types, which seems incompatible with your claim.

AFAICT that just means these objects can have padding, and `memcpy` can copy padding bytes just fine. There is no other option, since as mentioned above, [basic.types.general.4.1](http://eel.is/c++draft/basic.types.general#4.sentence-1) guarantees that all objects occupy a _sequence_ of `sizeof(T)` `unsigned char`s.

> I am not convinced that is well-defined; can you provide me with quotes of the normative text that define the behavior?

* [basic.types.general.4.1](http://eel.is/c++draft/basic.types.general#4.sentence-1) guarantees that all objects occupy a _sequence_ of `sizeof(T)` `unsigned char`s.
* [[cstring.syn.1]](http://eel.is/c++draft/cstring.syn#1 defers to [7.24.2.1](https://port70.net/~nsz/c/c11/n1570.html#7.24.2.1), which specifies `std::memcpy`, which just "copies n _characters_" from one object to another (in this case, from a `NonTrivial[2]` to a `unsigned char[sizeof(NonTrivial[2])];`. 

I don't think there is anything else to define: memcpy copies characters of the representation, objects have representations, hence you can copy from one representation to another. 

C++ adds lifetimes, so when you copy into a new representation, the old one whose storage is being reused is dead (which is fine for `unsigned char`, it does not have destructors), and a new object is created (which is also fine for `unsigned char`, since it is trivially copyable and constructible).

Reply:
I think this example can be further reduced to whether: 

```c++
NonTrivial foo{};
unsigned char c = *((unsigned char*)&foo);
```
is UB or not since this is essentially what memcpy semantically does (it then writes `c` somewhere else; nothing here technically prevents preserving provenance).

Reply:
See http://wg21.link/p18392r2 , which hints that there is unanimous consensus that accessing the entire object representation through a char-like type for contiguous objects should be well-defined and fixed as DR (defect report?).

It forwards to http://wg21.link/p1945 which address non-contiguous types (virtual base sub-objects), but in this particular case, the type in the example is contiguous, or the example can be easily fixed to only run if it is contiguous.

Reply:
I don't think that's the same issue, but I'm starting to lean towards your reading of the words for std::memcpy, even though the 9899 definition is somewhat problematic given how 14882 defines objects.

Having conceded that though, I still believe that making such calls should be UB for memcpy_async. There's little you can do with the resulting memory location; if the type is standard layout, you could extract a subobject's offset and read its value if it's trivially copyable, but then you can just copy the subobject instead. (Non trivially copyable objects don't have values per the standardese definition of the word.)

Is there any particular use cases that are stopped by not making the calls with non trivially copyable types valid for memcpy_async? I'd rather we build in some safety into the C-like APIs we provide unless there's some compelling reason to do so.

Reply:
> I'd rather we build in some safety into the C-like APIs we provide unless there's some compelling reason to do so.

Making this "library UB" or not is a design trade-off. Is the value proposition for making this library UB worth it? AFAICT, the costs of this trade-offs are:

- how do we teach it: yet another case of UB, need to mention it when mentioning `memcpy_async` instead of just saying "exact same semantics as `memcpy`
- can we catch it and how: I don't think we can catch it at the library level (implementing this as `memcpy_async<T>(T*,T*,size_t)` and failing on T=NonTriviallyCopyable has a bunch of issues), catching this at the compiler level would be hardish and not worth the effort, catching this at the sanitizer level would be very hard, so I think this puts this in the bag of UB that we can't diagnose.
- complicates our docs: it would be simpler to just say "exact same semantics as `memcpy`", whatever those might be, and if a DR lands that clarifies the behavior that all existing compilers implement, that automatically would apply.

and the value of making this UB is: 

- optimizations: if we make this an intrinsic (which we probably should if it isn't already), are there any optimizations that this allows beyond the tautological  "dead code removal because this would be UB" ?

What other value those it add? 

---

I think the simplest, future-proof thing to do here is to just say that this has the exact same semantics as `memcpy`. It doesn't help users much, because the semantics of `memcpy` aren't 100% clear in the standard, but that's the standard fault, and it is easier to solve that via a defect report or a paper, than try to pick one or the other side here.

Reply:
Well, memcpy_async is supposed to be a replacement for assigning from global memory to shared memory; if we allow byte-wise copies to be done for non trivially copyable types, we'll enable a class of bugs where users blindly do the recommended thing and get invalid behavior. By diagnosing the cases that are suspect (as the implementation does right now), we eliminate a subset of that class of bugs.

@hcedwar can you weigh in here from the point of view of platform exposure?

Reply:
We don't want to fall back to magically having our semantics changed from under us when a DR is applied to the standard. I really want to be able to vet any DR'd changes against the intent of our platform exposure for the parts of our stack outside of `cuda::std::`.

Remember that unlike memcpy, memcpy_async isn't merely a tool to punch holes in the object model; it's a hardware feature exposure mechanism. It needs to primarily serve as such, instead of being forced into the shape of a standard tool that is not serving such a goal.

Reply:
> We don't want to fall back to magically having our semantics changed from under us when a DR is applied to the standard. 

I would actually want that. The alternative is to live with very subtle differences in semantics between C++ and CUDA C++. That would add a significant teaching, documentation, and specification cost, as well as confusion, which would lead to bugs. 

> By diagnosing the cases that are suspect (as the implementation does right now),

As mentioned, this introduces a bunch of issues. Some examples are "violating our own spec for the signature of `memcpy_async`" (which isn't implemented as we specify it anymore), "breaking SFINAE for `memcpy_async`" (e.g. [this example](https://gcc.godbolt.org/z/cGs8hW) never copies a non-trivially copyable type with `memcpy_async`, and according to our spec it should compile, but it doesn't, because our implementation is non-conforming). There are probably more consequences of doing this.

> we'll enable a class of bugs where users blindly do the recommended thing and 
get invalid behavior. 
>
>  Remember that unlike memcpy, memcpy_async isn't merely a tool to punch holes in the object model; it's a hardware feature exposure mechanism. 

The C++ standard delegates the specification of `memcpy` to the C standard, in which it is essentially a tool for copying raw bytes, which is quite close to what the hardware does (the hardware doesn't care much about the bytes being trivially-copyable). 

It appears that what you want is to go beyond that, and have a safer API. In C++, that API is `std::copy`. An async version of `std::copy` that does the right thing depending on TriviallyCopyable, ContiguousRange, etc. can probably be implemented on top of a raw `memcpy_async` API.  

The only reason my original example would be "buggy" in CUDA C++ using `memcpy_async` is because we say that it exhibits "library UB". In C++, that non-trivially copyable type has contiguous storage on all implementations, and a paranoid user could write code to test that, which means that the code would not be buggy. Removing "library UB" from the `async_memcpy` specification would remove those bugs. 

I agree that a safer asynchronous copy API makes sense (e.g. `std::copy_async`). I don't know whether the value proposition of trying to bolt a degree of safety onto a `memcpy`-like API does (it looks to me that we are trading off catching some bugs by introducing others).

Reply:
> I would actually want that. The alternative is to live with very subtle differences in semantics between C++ and CUDA C++. That would add a significant teaching, documentation, and specification cost, as well as confusion, which would lead to bugs.

That's a fine stance to have, and you are free to have that stance, but that is not the stance _we_ have for functionalities that are not direct extensions of C++ APIs.

> As mentioned, this introduces a bunch of issues. Some examples are "violating our own spec for the signature of memcpy_async" (which isn't implemented as we specify it anymore),

I don't understand what this means. How do we break the spec? Compile-time UB is still a thing, failing to compile *is* a valid way to handle UB; see [[defns.undefined]](https://eel.is/c++draft/defns.undefined). Please elaborate on how you think we break the spec.

> "breaking SFINAE for memcpy_async" (e.g. this example never copies a non-trivially copyable type with memcpy_async, and according to our spec it should compile, but it doesn't, because our implementation is non-conforming).

Where does our spec say it should compile? The spec presents overloads that accept `void *` arguments and nowhere promises any templated overloads that you can SFINAE on. The example you've linked is not valid, because it tries to SFINAE on a function call that is not specified to give you SFINAE. What part of the docs made you think the example is valid? It's a problem with that part of the spec, not with this.

> The C++ standard delegates the specification of memcpy to the C standard, in which it is essentially a tool for copying raw bytes, which is quite close to what the hardware does (the hardware doesn't care much about the bytes being trivially-copyable).

That's fine, and if/when we expose `cuda::std::memcpy`, we will follow that specification. `memcpy_async` is a separate feature with its own design that's _based_ on `std::memcpy`, and therefore has its own specification.

> The only reason my original example would be "buggy" in CUDA C++ using memcpy_async is because we say that it exhibits "library UB". In C++, that non-trivially copyable type has contiguous storage on all implementations, and a paranoid user could write code to test that, which means that the code would not be buggy. Removing "library UB" from the async_memcpy specification would remove those bugs.

Your earlier example is not driving this decision; it's the original use case for hardware exposure that is driving this decision. Consider:

```cuda
template<typename T>
__global__ void kernel(T * global_mem, std::size_t num_elements) {
    __shared__ T shared_mem[THREAD_COUNT];
    for (auto i = 0; i < num_elements / THREAD_COUNT; ++i) {
        shared_mem[threadIdx.x] = global_mem[blockDim.x * i + threadIdx.x];
        // ... do computations
    }
    // ... handle the tail of the data set
}
```

`memcpy_async` is meant _primarily_ to enable the user to rewrite functions like this. Most functions like this will probably just expect to work on integral or floating-point types, which are trivially copyable. But suppose that it's meant to work with a multitude of types, and the user rewrites (as our messanging says they should!) the `shared_mem[threadIdx.x] = global_mem[blockDim.X * i + threadIdx.x];` line with a call to memcpy_async (probably in the vein of `cuda::memcpy_async(cg::this_thread_block(), shared_mem, global_mem[blockIdx.x * i], sizeof(T), pipeline);`. This is _good_, but if we allowed non-trivially-copyable types, this may introduce a _very subtle_ bug of doing a bytewise copy of a type with custom copy semantics, and the rewrite has now changed the semantics of the program in question... ...despite it being a rewrite that we recommend users do. Whoopsie. Not a great look. If this fails to compile with a very explicit error message (except on GCC 4.8, but that can't be helped), we've saved the user from rewriting a kernel into a form that breaks its semantics.

If, however, the user wants to still force it... they can just cast to `void *` or `std::byte *` (and wrap the size in `cuda::aligned_size_t`).

If we ever arrive in a place where there's more _actual_ use-cases for `memcpy_async` than the above rewrite, I'll happily come back to this discussion and possibly revisit it. After all, removing a requirement like this from a function is a backwards compatible change, so when circumstances of `memcpy_async` change, we can very easily go in this direction if it is deemed reasonable.

Reply:
> Where does our spec say it should compile? The spec presents overloads that accept void * arguments and nowhere promises any templated overloads that you can SFINAE on. 

The spec only provides `void*` overloads. If the implementation would match the spec, my _valid_ SFINAE example that never invokes UB even according to our restrictive spec would compile just fine. The reason it doesn't is because we provide more overloads than what the spec provides. 

> If, however, the user wants to still force it... they can just cast to void * or std::byte * (and wrap the size in cuda::aligned_size_t).

[My example](https://gcc.godbolt.org/z/cGs8hW) is well-defined according to our spec (it never calls `memcpy_async` with a non-trivially copyable type), yet we reject it. Having to use `void*` here is IMO a bug in the implementation.

Other examples that call `memcpy_async` with a non-trivially copyable type are UB according to our spec. We shouldn't be recommending to use `void*` to workaround that.

---

Adding "safety" to `memcpy` is a trade-off. I think that the costs of doing this are (at least):

- deviating from standard C++ `memcpy` semantics, 
- adding "library UB" that's impossible to check for properly (we can't check `void*`), 
- deviating from our documented API, and
- rejecting a subset of valid programs.

I don't think these costs are worth the value of catching the subset of bugs mentioned. This is a value proposition, if you and others think that these costs are worth their value, I respect that. 

C++ solution to this problem is `std::copy`, which is implemented on top of the "dangerous" `std::memcpy`. I think it would be better to have a safe `copy_async` API that's actually safe to use, instead of paying a high cost to get `memcpy_async` only half-way there.

Reply:
> The spec only provides void* overloads. If the implementation would match the spec, my valid SFINAE example i that never invokes UB even according to our restrictive spec would compile and execute just fine. The reason it doesn't is because we provide more overloads than what the spec provides.

Regarding "more overloads": since `memcpy_async` is not an [_addressable function_](http://eel.is/c++draft/library#namespace.std-6), you are not allowed to take its address, which makes it impossible for you to tell what overloads are present in the implementation and whether they exactly match the overloads in the API surface or not. The spec defines the API; the implementation implements that API (the total sum of the overloads present in the implementation is the full API surface, and there's no ambiguitites between them in valid calls to the API surface).

> My example is well-defined according to our spec (it never calls memcpy_async with a non-trivially copyable type), yet we reject it.

It's very interesting that you think it does not call memcpy_async with a non trivially copyable type, because the diagnostic you are getting comes from a line in the program that invokes memcpy_async with a non trivially copyable type. The call on line 39 is unconditional; the "trait" you defined is unconditionally true; that means that line 39 unconditionally leads to executing line 22, which is undefined behavior per the spec. Unless of course you want to argue that this is only true if `foo` is ever invoked, but that's playing semantics for no useful reason.


> deviating from standard C++ memcpy semantics, etc.

We necessarily deviate from memcpy semantics.

> adding "library UB" that's impossible to check for properly (we can't check void*),

It is possible to check for the trap in the intended and suggested use cases.

> deviating from our documented API,

Citation still needed. Repeating the assertion that your example is well-defined according to the spec does not make it so.

> rejecting a subset of valid programs. and

Ditto.

> higher teachability, documentation, and implementation costs due to all of the above.

This is highly subjective and I don't think I agree with any of these.


--------------------------


I appreciate the feedback, and I also appreciate the course this discussion has run; I think I am now closer to agreeing with you about how well defined `std::memcpy` is for a certain class of calls (though I'm pretty sure I'm not there yet, at least not fully ;)). If/when the intented use cases of `memcpy_async` change, we will revisit the conscious decisions made about the requirements placed on calls to `memcpy_async`. For now, since I feel like we have reached a point in the discussion where we are bound to run in circles, I'll resolve this conversation, in the interests of keeping the PR view clearer than it has become thanks to the volume of this thread.

Reply:
> It's very interesting that you think it does not call memcpy_async with a non trivially copyable type, because the diagnostic you are getting comes from a line in the program that invokes memcpy_async with a non trivially copyable type. 

I understand that this is hard to see in that example. [This program shows the crux of the issue (godbolt)](https://gcc.godbolt.org/z/W47Yxx)):  

```c++
template <typename T>
__global__ void kernel(...) {
    if (std::is_trivially_copyable_v<T>) {
        cuda::memcpy_async(g, shared, global, sizeof(NTC), b);
    } else {
        shared[i] = global[i];
    } 
}
```

I believe this program is valid according to our specification and should be accepted. Yet our current implementation rejects it. 

This program never calls `memcpy_async` with a non-trivially copyable type and does not have UB. Yet the compilation error rejects it under that premise. 

I hope this example makes the issue clear.

Reply:
That's a better example, thanks; but getting rid of a diagnostic in that case is also simple: use if constexpr instead.

I've heard the solutions you've been proposing for several comments. We will relitigate if/when the intended use cases for memcpy_async change.

Reply:
> That's a better example, thanks; but getting rid of a diagnostic in that case is also simple: use if constexpr instead.

The issue remains that a valid program is rejected. If this program is not valid, the docs should IMO say so.

Also, replacing SFINAE with `if constexpr` isn't trivial, so I don't think that's a good solution.

Reply:
Okay, that is a fair point. I wish you could've started with examples like that :)

@c0riolis can you split the second part of the last paragraph off into its own remark, and reword it to something like, "a program that contains a call to `memcpy_async` with arguments pointing to objects of types that are not TriviallyCopyable is ill-formed, no diagnostic required."?

Reply:
> I wish you could've started with examples like that :)

Apologies, really. There are a couple of issues mixed in this thread, and github is not great at keeping different threads of discussion separate (or cleaning them up afterwards). 

I didn't know that it was intentional to deviate from `std::memcpy` here, so we had to resolve that, and then it turned out that I disagree with that design. For me that's not an issue, but raising the tradeoffs I see and understanding the ones that others have raised is. In retrospect, a github thread is not the best avenue for that.

Reply:
Yeah, that sentiment is a part of what I meant. I didn't intend that as an accusation of any sort ;)

Reply:
> @c0riolis can you split the second part of the last paragraph off into its own remark, and reword it to something like, "a program that contains a call to memcpy_async with arguments pointing to objects of types that are not TriviallyCopyable is ill-formed, no diagnostic required."?
Done.